primitives/consensus/sassafras/src/ticket.rs

```diff
@@ -0,0 +1,149 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Primitives related to tickets.
+
+use crate::{Randomness, RingVrfSignature, VrfInput, VrfOutput, VrfSignData, SASSAFRAS_ENGINE_ID};
+use scale_codec::{Decode, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+use sp_consensus_slots::Slot;
+use sp_std::vec::Vec;
+
+pub use sp_core::ed25519::{Public as EphemeralPublic, Signature as EphemeralSignature};
+
+/// Ticket identifier.
+///
+/// Its value is the output of a VRF whose inputs cannot be controlled by the
+/// creator of the ticket (refer to [`ticket_id_vrf_input`] parameters).
+/// Because of this, it is also used as the ticket score to compare against
+/// the epoch ticket's threshold.
+pub type TicketId = u128;
+
+/// Ticket data persisted on-chain.
+#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode, MaxEncodedLen, TypeInfo)]
+pub struct TicketBody {
+	/// Attempt index.
+	pub attempt_idx: u32,
+	/// Ed25519 ephemeral public key representing ticket ownersip.
+	/// (i.e. whoever has the secret, is the owner)
+	pub erased_public: EphemeralPublic,
```

Comment:
(In the meantime I've added `revealed_public` as well 😃)

The naming has to do with an additional functionality these keys have beside being used to claim the ticket.

The additional utility is "arbitrary data encryption" so that only the ticket owner is able to decrypt it.
- `erased_public`: is valid until the ticket is not claimed/used. Once validator identity is disclosed, this key is "erased". I.e. invalidated. (e.g. can be used by someone who - for some reason - knows the identity of a future validator, and maybe encrypted transactions?).  This is "erased" to guarantee what is called *forward security*. I.e. Once I reveal my identity, even if my secret keys are exposed in the future these are not valid (so that the key can't be used to make stuff in the past).
- `revealed_public`: becomes valid after the ticket is claimed.

Reply:
> for some reason - knows the identity of a future validator

Are valid reasons known now? Only if the validator leaks this info?

Reply:
Not that I'm aware of as are not directly linked to the protocol. 
And yes, only if the validator leaks this info